### DISCUSSION TITLE
Make Argon2 hashing options (memory/time cost, etc configurable.)

### DISCUSSION THREAD

**Discussion by TonyLovesDevOps** [directus/directus#7670] [Feature Requests]
Hi there,

For compliance/security purposes, we need to modify the options passed to the `argon2.hash` function for hashing user passwords (see options [on the node-argon2 wiki](https://github.com/ranisalt/node-argon2/wiki/Options)).

For example, instead of `const passwordHashed = await argon2.hash(password);` we need to do something like:

```javascript
const passwordHashed = await argon2.hash(password, {
    memoryCost: 196608, // 192MiB instead of the default of 4MiB
    timeCost: 10 // ten passes instead of the default of three
}
```

I _think_ the only places this would apply would be:
[acceptInvite in user.ts](https://github.com/directus/directus/blob/7dfc5dc6afa342909179b2835c7fccb4f300a045/api/src/services/users.ts#L342)
[resetPassword in user.ts](https://github.com/directus/directus/blob/7dfc5dc6afa342909179b2835c7fccb4f300a045/api/src/services/users.ts#L406)
[in usersPasswd in passwd.ts for the CLI](https://github.com/directus/directus/blob/d64ca14348184e69e4348c8c61bbf88ffcc9937a/api/src/cli/commands/users/passwd.ts#L17)
[in init in index.ts](https://github.com/directus/directus/blob/d64ca14348184e69e4348c8c61bbf88ffcc9937a/api/src/cli/commands/init/index.ts#L98)

I'm not sure about [the other calls to argon2.hash](https://github.com/directus/directus/search?q=argon2.hash) that don't appear to be password-related.

Questions:
1. Is there a way to accomplish this today that I'm not thinking of, with extensions or some other built-in mechanism?
2. If not, would you be amenable to a PR adding this? And if so, do you have any preferences/examples of how to implement a configuration setting like this?
3. Alternatively, should we fork the code base and change it for ourselves and build our own docker image (we'd really like to avoid that as a long-term strategy)?

In the interest of full disclosure on question 2 offering to submit a PR: I'm sort of a John Rambo when writing in node.js: I can probably get the job done, but there will be consequences. 💥

**Comment by rijkvanzanten** [comment]
Heya!

1) Nope!

2) Yes!

> And if so, do you have any preferences/examples of how to implement a configuration setting like this?

Ideally, we expose the full argon options object through environment variables. We already have a util function that auto-converts the env vars into a camelCased typed object, so that would be perfect for this. As a prefix, we could use `HASH_` (so `HASH_MEMORY_COST` -> `memoryCost` in the options object.

```js
const argonConfig = getConfigFromEnv('HASH_');
await argon2.hash(password, argonConfig);
```

3) That's a last resort, but I'd strongly recommend against if at all possible 👍🏻 

> In the interest of full disclosure on question 2 offering to submit a PR: I'm sort of a John Rambo when writing in node.js: I can probably get the job done, but there will be consequences. 💥

Hehe, that's where I come in with the reviews on PRs 😉 By utilizing the existing util, the actual implementation should be fairly straightforward I think. The main thing that's easy to forget is the documentation.

**Reply by TonyLovesDevOps** [reply]
Great, thanks for the pointers. I'll start putting together a PR!

**Reply by TonyLovesDevOps** [reply]
> Ideally, we expose the full argon options object through environment variables.

Do we really want *all* of the options exposed, for example [raw](https://github.com/ranisalt/node-argon2/wiki/Options#raw) which returns the raw hash (before base64 encoding it) as a `Buffer` object? What happens if `passwordHashed` is a `Buffer` here?
```javascript
await this.knex('directus_users').update({ password: passwordHashed, status: 'active' }).where({ id: user.id });
```

I'm also wary about the value/danger of exposing `version`, `salt`, `saltLength`, and `associatedData` -- especially `salt` seems like a footgun.

**Reply by rijkvanzanten** [reply]
Good points. Normally, I'd err on the side of flexibility, but I agree this really opens up some footguns. I just know that if we allow a few, but not all, options, that others will start requesting more and more options until we've come full circle anyways 🤔

I do only expect power users that have a specific requirement (just like yourself) to start modifying these specific values, so maybe it's not a problem as long as we add a "please make sure what you're doing" warning to the reference

**Reply by TonyLovesDevOps** [reply]
I defer to your experience with this. I've gotten all of the options working except for setting `raw` to true, which results in this:

```
mysql> select * from directus.directus_users\G
   ...
   password: {"type":"Buffer","data":[50,73,117,41,165,114,106,74,36,203,34,91,86,100,84,74]}
   ...
 ```
🙄 which throws an error when the user tries to log in and `argon2.verify` gets called.

I really don't know how to handle `raw` in a sane way, since even stringifying it doesn't store any of the other required parameters that the `argon2.verify` function expects (e.g. everything up to the last part of `$argon2i$v=19$m=4096,t=14,p=1$w4bC3AxERB4ODSE5YfQZGg$<base64 encoded output of what setting raw to true returns>`).

What do you think is sane to do here?

**Reply by rijkvanzanten** [reply]
The way we use argon2 right now, I don't think there's every going to be a use for the `raw` option 🤔 

I think we should ignore it by excluding that config flag (configFromEnv util has a parameter to ignore specific keys)

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
